source/fundamentals/data-formats/documents.txt

```diff
@@ -14,34 +14,351 @@ Documents
+       .append("novels", new BsonArray(Arrays.asList(
+           new BsonDocument().append("title", new BsonString("One Hundred Years of Solitude")).append("yearPublished", new BsonInt32(1967)),
+           new BsonDocument().append("title", new BsonString("Chronicle of a Death Foretold")).append("yearPublished", new BsonInt32(1981)),
+           new BsonDocument().append("title", new BsonString("Love in the Time of Cholera")).append("yearPublished", new BsonInt32(1985))
+      )));
+
+You can insert this document into a collection using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   // MongoClient mongoClient = ...
+
+   MongoCollection<BsonDocument> collection = database.getCollection("authors", BsonDocument.class);
+   MongoDatabase database = mongoClient.getDatabase("fundamentals_data");
+
+   InsertOneResult result = collection.insertOne(author);
+
+Once you perform a successful insert, you can retrieve the sample document
+data from the collection using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   import com.mongodb.client.model.Filters;
+   ...
+
+   BsonDocument doc = collection.find(Filters.eq("novels.yearPublished", 1967)).first();
+   if (doc != null) {
+       System.out.println("_id: " + doc.getObjectId("_id").getValue()
+               + ", name: " + doc.getString("name").getValue()
+               + ", lastUpdated: " + new Date(doc.getDateTime("lastUpdated").getValue()));
+
+       doc.getArray("novels").forEach((novel) -> {
+           System.out.println("title: " + novel.asDocument().getString("title").getValue()
+                   + ", yearPublished: " + novel.asDocument().getInt32("yearPublished").getValue());
+       });
+   }
+
+The output should resemble the following:
+
+.. code-block:: none
+   :copyable: false
+
+   _id: 5fb5fad05f734e3794741a35, name: Gabriel García Márquez, lastUpdated: Thu Mar 26 00:00:00 EDT 2020
+   title: One Hundred Years of Solitude, yearPublished: 1967
+   title: Chronicle of a Death Foretold, yearPublished: 1981
+   title: Love in the Time of Cholera, yearPublished: 1985
+
+For more information on retrieving and manipulating MongoDB data, see our
+:doc:`CRUD guide </fundamentals/crud>`.
+
+For more information on this class, see the
+:java-docs:`BsonDocument API documentation </apidocs/bson/org/bson/BsonDocument.html>`.
 
 BasicDBObject
 -------------
 
-- Description
-- Code example of declaring one and appending fields
-- Link to Fundamentals > CRUD for more information
-- Link to API documentation
+The ``BasicDBObject`` class allows you to access and manipulate document data
+using Java types. We recommend that you avoid using this class unless you
+are migrating an application from that uses an older driver version. In
+addition to deprecation, this class has the following limitations:
+
+- This class implements the ``DBObject`` interface, so you cannot extend its
+  API without breaking binary compatibility.
```

Comment:
Perhaps I'm just not familiar with "binary compatibility" or API extension constraints, but it's unclear to me how the first clause relates to the second clause (I'm assuming a causal relationship). It could also be helpful to define "binary compatibility" here (are we talking about binaries as in programs? Or binary representation of objects in process memory?

Reply:
I copied this from the existing docs, but I'm assuming it is referring to https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_packages -- see example 4. (A more recent doc: https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html )

As I understand it, "extend its API" means add a method to a class that implements it such as BasicDBObject. Suppose you had a compiled app that uses driver v4.1, but v4.2 came out with a new method, BasicDBObject#coolNewMethod. Normally, you could just upgrade the jar in the classpath and restart the app (or hot-reload) and it would work fine.  However, since you broke binary compatibility with v4.2 by adding a method, you need to recompile your app for it to work with v4.2. However, if there was some non-breaking change such as a private method change, your app would continue to work without recompilation because it's not visible to your binary.

Admittedly, it is a little unclear which object the pronoun "its" refers to so I can reword. It's also probably less frequently encountered than the next item, so I'll re-order it in the list.